src/bridge/SequencerInbox.sol

```diff
@@ -356,24 +538,139 @@ contract SequencerInbox is DelegateCallAware, GasRefundEnabled, ISequencerInbox
+            headerByte == ZERO_HEAVY_MESSAGE_HEADER_FLAG;
+    }
+
+    /// @dev    Form a hash of the data taken from the calldata
+    /// @param  data The calldata to be hashed
+    /// @param  afterDelayedMessagesRead The delayed messages count read up to
+    /// @return The data hash
+    /// @return The timebounds within which the message should be processed
+    function formCallDataHash(bytes calldata data, uint256 afterDelayedMessagesRead)
         internal
         view
-        returns (bytes32, TimeBounds memory)
+        returns (bytes32, IBridge.TimeBounds memory)
     {
-        (bytes memory header, TimeBounds memory timeBounds) = packHeader(afterDelayedMessagesRead);
-        return (keccak256(header), timeBounds);
+        uint256 fullDataLen = HEADER_LENGTH + data.length;
+        if (fullDataLen > maxDataSize) revert DataTooLarge(fullDataLen, maxDataSize);
+
+        (bytes memory header, IBridge.TimeBounds memory timeBounds) = packHeader(
+            afterDelayedMessagesRead
+        );
+
+        // the batch poster is allowed to submit an empty batch, they can use this to progress the
+        // delayed inbox without providing extra batch data
+        if (data.length > 0) {
+            // The first data byte cannot be the same as any that have been set via other methods (eg 4844 blob header) as this
+            // would allow the supplier of the data to spoof an incorrect 4844 data batch
+            if (!isValidCallDataFlag(data[0])) revert InvalidHeaderFlag(data[0]);
+
+            // the first byte is used to identify the type of batch data
+            // das batches expect to have the type byte set, followed by the keyset (so they should have at least 33 bytes)
+            // if invalid data is supplied here the state transition function will process it as an empty block
+            // however we can provide a nice additional check here for the batch poster
+            if (data[0] & DAS_MESSAGE_HEADER_FLAG != 0 && data.length >= 33) {
+                // we skip the first byte, then read the next 32 bytes for the keyset
+                bytes32 dasKeysetHash = bytes32(data[1:33]);
+                if (!dasKeySetInfo[dasKeysetHash].isValidKeyset) revert NoSuchKeyset(dasKeysetHash);
+            }
+        }
+        return (keccak256(bytes.concat(header, data)), timeBounds);
+    }
+
+    /// @dev   Form a hash of the data being provided in 4844 data blobs
+    /// @param afterDelayedMessagesRead The delayed messages count read up to
+    /// @return The data hash
+    /// @return The timebounds within which the message should be processed
+    function formBlobDataHash(uint256 afterDelayedMessagesRead)
+        internal
+        view
+        returns (
+            bytes32,
+            IBridge.TimeBounds memory,
+            uint256
+        )
+    {
+        bytes32[] memory dataHashes = reader4844.getDataHashes();
+        if (dataHashes.length == 0) revert MissingDataHashes();
```

Comment:
No change necessary, I just wanted to point out that this is a great check and we should keep it. Specifically, I'm worried that some RPC providers that aren't ready for 4844 yet might drop the blobVersionedHashes field from eth_estimateGas and would otherwise return too low of an estimate. Reverting here prevents that from happening.